### PR TITLE
Core: New formulation of linear and non-linear residuals

### PR DESCRIPTION
**Pull request by tduretz**
* I have changed the formulation of the finite difference residuals. This is an important work since it affects residual evaluations and the finite difference coefficients of both the Picard (Stokes)  and the Jacobian matrices. 
See details in Python_CodeGen/AssembleGeneralStiffness_MDOODZ_7.0.ipynb file.

* I have been debugging the code using valgrind and killed bugs which should solve the problems of William.

* So far the tests that passed (not yet in an automatic way) are: VEP_Duretz18.txt, TopoBenchCase1.txt, LithosScale.txt and looked reliable with Setup03_strongCircleAnisoConst.txt

### CODE REVIEW

**Review by halterw — APPROVED** [review]

File: SOURCE/GridRoutines.c [hunk 1]
```diff
@@ -105,71 +105,79 @@ void ExpandCentroidArray( double* CentroidArray, double* temp, grid* mesh, param
...
+    int*     temp_flag = DoodzCalloc((Ncx+2)*(Ncz+2), sizeof(int));
     
     // Fill interior points
     for (k=0; k<Ncx; k++) {
         for (l=0; l<Ncz; l++) {
             c0 = k + l*(Ncx);
             c1 = k + (l+1)*(Ncx+2) + 1;
-            temp[c1] = CentroidArray[c0];
+            temp[c1]      = CentroidArray[c0];
+            temp_flag[c1] = mesh->BCp.type[c0];
```
Comment (kulakovri) [inline SOURCE/GridRoutines.c:126]: Is "temp" - "temporary" or "temperature"?

File: SOURCE/Main_DOODZ.c [hunk 2]
```diff
@@ -144,243 +132,198 @@ int RunMDOODZ( int nargs, char *args[] ) {
...
-                // Marker chain polynomial fit
-                MarkerChainPolyFit( &topo, &topo_chain, model, mesh );
-                
-                // Call cell flagging routine for free surface calculations
-                CellFlagging( &mesh, model, topo, scaling );
-            }
-            // Set particles coordinates
-            PutPartInBox( &particles, &mesh, model, topo, scaling );
+            // Project topography on vertices
+            ProjectTopography( &topo, &topo_chain, model, mesh, scaling, mesh.xg_coord, 0 );
```
Comment (kulakovri) [inline SOURCE/Main_DOODZ.c:151]: What's the difference between "ProjectTopography" and "BuildInitialTopography" methods?
Maybe we should reflect it somehow in name?

**Review by kulakovri — CHANGES_REQUESTED** [review]

File: SOURCE/GridRoutines.c (hunk 1 above)
Comment (tduretz) [reply SOURCE/GridRoutines.c:126]: temporary, I have renamed the ambiguous "temp" to "tmp"

File: SOURCE/Main_DOODZ.c (hunk 2 above)
Comment (tduretz) [reply SOURCE/Main_DOODZ.c:151]: "ProjectTopography" projects topography from the markers onto the grid.
"BuildInitialTopography" builds the initial topography (on the markers)
I've left a TODO for future refactoring.

**Review by kulakovri — APPROVED** [review]

**Pull request merged**